cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -132,4 +179,88 @@ public void create() throws IOException {
       LOG.info("Upgraded hbase table: {}", tableName);
     }
   }
+
+
+  // Moves dataset instance meta entries into new table (in system namespace)
+  // Also updates the spec's name ('cdap.user.foo' -> 'foo')
+  private void upgradeDatasetInstanceMDS() throws IOException, DatasetManagementException, InterruptedException,
+    TransactionFailureException {
+    LOG.info("Upgrading dataset instance mds.");
+    datasetInstanceMds.execute(new TransactionExecutor.Function<UpgradeMdsStores<DatasetInstanceMDS>, Void>() {
+      @Override
+      public Void apply(UpgradeMdsStores<DatasetInstanceMDS> ctx) throws Exception {
+        MDSKey key = new MDSKey(Bytes.toBytes(DatasetInstanceMDS.INSTANCE_PREFIX));
+        DatasetInstanceMDS newMds = ctx.getNewMds();
+        Map<MDSKey, DatasetSpecification> dsSpecs = ctx.getOldMds().listKV(key, DatasetSpecification.class);
```

Comment:
is this correct? should you be reading ``MDSKey`` from the old MDS? or am I missing something?